homeassistant/components/imap/coordinator.py

```diff
@@ -100,5 +102,7 @@ async def async_wait_server_push(self) -> None:
     async def shutdown(self, *_) -> None:
         """Close resources."""
         if self.imap_client:
+            if self.imap_client.has_pending_idle():
```

Comment:
This is needed to cancel any waiting for server push. Without it, it will throw an exception like below:

```log
2023-03-15 14:59:08.643 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved:   File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/vscode/.vscode-server/extensions/ms-python.python-2023.4.1/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/home/vscode/.vscode-server/extensions/ms-python.python-2023.4.1/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/home/vscode/.vscode-server/extensions/ms-python.python-2023.4.1/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 317, in run_module
    run_module_as_main(options.target, alter_argv=True)
  File "/home/vscode/.vscode-server/extensions/ms-python.python-2023.4.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 238, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/vscode/.vscode-server/extensions/ms-python.python-2023.4.1/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/workspaces/core/homeassistant/__main__.py", line 226, in <module>
    sys.exit(main())
  File "/workspaces/core/homeassistant/__main__.py", line 214, in main
    exit_code = runner.run(runtime_conf)
  File "/workspaces/core/homeassistant/runner.py", line 179, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 633, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 600, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 1888, in _run_once
    handle._run()
  File "/usr/local/lib/python3.10/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/workspaces/core/homeassistant/core.py", line 1114, in _onetime_listener
    self._hass.async_run_job(listener, event)
  File "/workspaces/core/homeassistant/core.py", line 641, in async_run_job
    return self.async_run_hass_job(HassJob(target), *args)
  File "/workspaces/core/homeassistant/core.py", line 596, in async_run_hass_job
    return self.async_add_hass_job(hassjob, *args)
  File "/workspaces/core/homeassistant/core.py", line 492, in async_add_hass_job
    task = self.loop.create_task(hassjob.target(*args), name=hassjob.name)
Traceback (most recent call last):
  File "/home/vscode/.local/lib/python3.10/site-packages/aioimaplib/aioimaplib.py", line 301, in wrapper
    res = await coro(self, *args, **kargs)
  File "/home/vscode/.local/lib/python3.10/site-packages/aioimaplib/aioimaplib.py", line 464, in logout
    response = (await self.execute(Command('LOGOUT', self.new_tag(), loop=self.loop)))
  File "/home/vscode/.local/lib/python3.10/site-packages/aioimaplib/aioimaplib.py", line 414, in execute
    await self.pending_sync_command.wait()
  File "/home/vscode/.local/lib/python3.10/site-packages/aioimaplib/aioimaplib.py", line 206, in wait
    await self._event.wait()
  File "/usr/local/lib/python3.10/asyncio/locks.py", line 214, in wait
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 456, in wait_for
    return fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/workspaces/core/homeassistant/components/imap/coordinator.py", line 106, in shutdown
    await self.imap_client.logout()
  File "/home/vscode/.local/lib/python3.10/site-packages/aioimaplib/aioimaplib.py", line 700, in logout
    return await asyncio.wait_for(self.protocol.logout(), self.timeout)
  File "/usr/local/lib/python3.10/asyncio/tasks.py", line 458, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError
```